Universal ZIP Creation

trieloff opened the pull request:
First step in creating a universal packaging and deployment tool

trieloff committed:
feat(archive): add multicloud wrapper to archive

trieloff committed:
feat(multicloud): add wrapper template

trieloff committed:
Merge remote-tracking branch 'origin/main' into universal-zip

trieloff committed:
test(universal): test universal packaging

trieloff commented:
Question for @stefan-guggisberg and @tripodsan: I'm using `node-fetch` as the provider of `Request` and `Response` implementations, should I switch to `helix-fetch`?

trieloff committed:
fix(deploy): force deployment as raw web action in openwhisk

the universal wrapper assumes all actions are raw web actions

BREAKING CHANGE: The options `--web-export` and `--raw-http` have been removed as all actions will now be deployed to OpenWhisk as raw web actions

trieloff committed:
test(integration): add integration test with real deployment on CI

stefan-guggisberg commented:
the `node-fetch` implementation of `Request` and `Response` is fine. It's light-weight. I wouldn't use the current `helix-fetch` (i.e. `fetch-h2`) implementation.

once i am finished with the `helix-fetch` re-write (basics done, see https://github.com/stefan-guggisberg/poly-fetch for current status) we can consider using those implementations.

trieloff committed:
refactor(test): extract test utils

trieloff committed:
refactor(test): preserve environment vars

trieloff committed:
ci(test): use xunit reporter

trieloff committed:
fix(openwhisk): fix wrapper entry point

trieloff committed:
ci(test): output file for xunit reporter

trieloff committed:
test(openwhisk): fix entry point

trieloff committed:
ci(tests): upload test results

tripodsan commented on src/action_builder.js:
```diff
@@ -587,7 +587,12 @@ module.exports = class ActionBuilder {
       archive.directory(path.resolve(this._cwd, `node_modules/${mod}`), `node_modules/${mod}`);
     });
+
+    archive.directory(path.resolve(__dirname, '../node_modules/node-fetch'), 'node_modules/node-fetch');
```
hmm...this should be included automatically if imported somewhere in the code....

trieloff replied:
The `index.js` wrapper is never processed by webpack and even if the action imports `node-fetch`, it won't get picked up, hence the brute-force approach.

tripodsan replied:
> even if the action imports `node-fetch`, it won't get picked up, hence the brute-force approach.

but this would be wrong then.....

tripodsan commented on src/action_builder.js:
```diff
@@ -587,7 +587,12 @@ module.exports = class ActionBuilder {
+
     archive.append(JSON.stringify(packageJson, null, '  '), { name: 'package.json' });
+    // universal serverless wrapper
+    archive.append(fse.readFileSync(path.resolve(__dirname, 'template', 'index.js')).toString(), { name: 'index.js' });
```
```suggestion
    archive.file(path.resolve(__dirname, 'template', 'index.js'), { name: 'index.js' });
```

tripodsan commented on src/cli.js:
```diff
@@ -86,16 +86,6 @@ class CLI {
         description: 'Specifies the action kind.',
         default: 'nodejs:10',
       })
-      .option('web-export', {
```
??? we still need this to differentiate between web actions and _private_ one.

tripodsan commented on src/cli.js:
```diff
@@ -86,16 +86,6 @@ class CLI {
-        type: 'boolean',
-        default: true,
-      })
-      .option('raw-http', {
```
for web-action, we could make raw the default, but it still needs to be configurable.

tripodsan commented on src/template/index.js:
```diff
@@ -0,0 +1,248 @@
+ * @param {string} type - content type
+ * @returns {boolean} {@code true} if content type is binary.
+ */
+function isBinary(type) {
```
this list it more comprehensive:
https://github.com/adobe/openwhisk-action-utils/blob/1c4000bb178a6c56159e53c965bd04837327e274/src/expressify.js#L107-L175

tripodsan commented on src/template/index.js:
```diff
@@ -0,0 +1,248 @@
+ * governing permissions and limitations under the License.
+ */
+/* eslint-disable no-param-reassign, no-underscore-dangle, import/no-extraneous-dependencies */
+const { Request } = require('node-fetch');
```
wouldn't it be better to have our own Request/Response abstraction?
I don't like the extra dependency to node-fetch.

what about creating a `node-fetch-api` package that contains the _empty_ classes ?

trieloff replied:
Yes, ideally something that just webpacks `Request` and `Response` from `node-fetch` and re-exports it. We can later switch transparently to `helix-fetch`

tripodsan commented on test/fixtures/simple/index.js:
```diff
@@ -9,8 +9,10 @@
  * OF ANY KIND, either express or implied. See the License for the specific language
  * governing permissions and limitations under the License.
  */
+const { Response } = require('node-fetch');
```
this should automatically include node-fetch to the modules..... so no need to add it extra.

this also shows the _drawback_ of the universal approach. every action needs an extra dependency to node-fetch (or to whatever provides the `Response`).

tripodsan reviewed (changes requested):
- maybe we should only enable the universal package with a `--universal` flag....
- removing web-export and raw will break a lot of actions
- the deploy step should handle all targets
- maybe the tool should be split in a `build` and a `deploy` tool.

trieloff committed:
ci(mocha): use private mocha version for now

trieloff committed:
ci(integration): run integration tests in a separate process

trieloff committed:
refactor(build): use archive.file

Co-authored-by: Tobias Bocanegra <tripodsan@users.noreply.github.com>

trieloff commented:
> - maybe we should only enable the universal package with a --universal flag....
> - removing web-export and raw will break a lot of actions

I think we should treat the new version as a separate package with a new name (as the `openwhisk` will be very misleading). Our own actions can then switch to the new tool once they have been refactored to use the universal API.

> - maybe the tool should be split in a build and a deploy tool.

I'm planning to split the deployment code into an `OpenWhiskDeployer` class.

> - the deploy step should handle all targets

Yes, my next step is to add a `LambdaDeployer` class

trieloff committed:
refactor(openwhisk): extract openwhisk deployer into separate class

trieloff committed:
refactor(openwhisk): wait for all delegated functions

trieloff committed:
Merge branch 'universal-zip' of github.com:adobe/openwhisk-action-builder into universal-zip

tripodsan commented:
@trieloff I think we should move this to another repository.... it makes no sense to keep it as a PR here.

trieloff commented:
Probably. But that means we need a new name and I hoped to procrastinate on that a bit more.

tripodsan commented:
> Probably. But that means we need a new name and I hoped to procrastinate on that a bit more.

`universal-serverless-builder`

trieloff committed:
feat(aws): add AWS deployer

wiy

trieloff committed:
feat(aws): upload file to temporary s3 bucket

trieloff committed:
feat(aws): deploy to lambda and create version tags

trieloff committed:
feat(aws): create api gateway with integration, routes and permissions

try at https://tu3y9a2ojl.execute-api.us-east-1.amazonaws.com/simple-package/simple-name_1.43

trieloff commented:
Now here: https://github.com/adobe/helix-deploy
